docs/architecture/adr-054-semver-compatible-modules.md

```diff
@@ -0,0 +1,408 @@
+with an `InterfaceRegistry` as they are now because there would be no way to retrieve them via the global registry.
+
+In order to simplify access to other modules using ADR 033, a public API module (maybe even one
+[remotely generated by Buf](https://docs.buf.build/bsr/remote-generation/go)) could be used by client modules instead
+of requiring to generate all client types internally.
+
+The big downsides of this approach are that it requires big changes to how people use protobuf types and would be a
+substantial rewrite of the protobuf code generator. This new generated code, however, could still be made compatible with
+the [`google.golang.org/protobuf/reflect/protoreflect`](https://pkg.go.dev/google.golang.org/protobuf/reflect/protoreflect)
+API in order to work with all standard golang protobuf tooling.
+
+It is possible that the naive approach of marshaling/unmarshaling in the ADR 033 router is an acceptable intermediate
+solution if the changes to the code generator are seen as too complex. However, since all modules would likely need
+to migrate to ADR 033 anyway with this approach, it might be better to do this all at once.
+
+### Approach C) Don't address these issues
+
+If the above solutions are seen as too complex, we can also decide not to do anything explicit to enable better module
+version compatibility, and break circular dependencies.
+
+In this case, when developers are confronted with the issues described above they can require dependencies to update in
+sync (what we do now) or attempt some ad-hoc potentially hacky solution.
+
+One approach is to ditch go semantic import versioning (SIV) altogether. Some people have commented that go's SIV
+(i.e. changing the import path to `foo/v2`, `foo/v3`, etc.) is too restrictive and that it should be optional. The
+golang maintainers disagree and only officially support semantic import versioning. We could, however, take the
+contrarian perspective and get more flexibility by using 0.x-based versioning basically forever. 
+
+Module version compatibility could then be achieved using go.mod replace directives to pin dependencies to specific
+compatible 0.x versions. For instance if we knew `foo` 0.2 and 0.3 were both compatible with `bar` 0.3 and 0.4, we
+could use replace directives in our go.mod to stick to the versions of `foo` and `bar` we want. This would work as
+long as the authors of `foo` and `bar` avoid incompatible breaking changes between these modules.
+
+Or, if developers choose to use semantic import versioning, they can attempt the naive solution described above
+and would also need to use special tags and replace directives to make sure that modules are pinned to the correct
+versions.
+
+Note, however, that all of these ad-hoc approaches, would be vulnerable to the minor version compatibility issues
+described above unless [unknown field filtering](./adr-020-protobuf-transaction-encoding.md#unknown-field-filtering)
+is properly addressed.
+
```

Comment:
Another downside of approach (c) is that no progress is made towards resolving cyclic module dependencies.